app/src/main/java/no/ntnu/klubbhuset/data/repository/VippsRepository.java

```diff
@@ -0,0 +1,110 @@
+import static no.ntnu.klubbhuset.util.PreferenceUtils.PREF_NO_FILE_FOUND;
+
+public class VippsRepository {
+    private static VippsRepository ourInstance;
+    private RequestQueue requestQueue;
+    private static final String TAG = "VippsRepository";
+    private Application context;
+
+
+    private Cache cache = Cache.getInstance();
+    private VippsCache vippsCache = VippsCache.getInstance();
+
+    public static VippsRepository getInstance(Application context) {
+        if (ourInstance == null) {
+            ourInstance = new VippsRepository(context);
+        }
+        return ourInstance;
+    }
+
+    private VippsRepository(Application context) {
+        this.requestQueue = Volley.newRequestQueue(context);
+        this.context = context;
+    }
+
+    /**
+     * Retrieves the vippsToken from Vipps, saves the value to the MutableLiveData and prefs
+     */
+    public LiveData<Resource<String>> getToken() {
+        String url = CommunicationConfig.getVippsTokenURL();
+        String prefToken = PreferenceUtils.getVippsToken(context.getApplicationContext());
+        MutableLiveData<Resource<String>> cached = vippsCache.getVippsToken();
+        if (!prefToken.equals(PREF_NO_FILE_FOUND)) {
+            // no token is found in prefs
+            loadVippsToken(cached, url);
+        } else if (tokenIsExpired(prefToken)) {
+            // token has expired
+            loadVippsToken(cached, url);
+        } else {
+            // prefs token is valid
+            cached.setValue(Resource.success(prefToken));
+        }
+        return cached;
+    }
+
+    private void loadVippsToken(MutableLiveData cached, String url) {
+        StringRequest request = new StringRequest(Request.Method.POST, url,
+                response -> {
+                    cached.setValue(Resource.success(response));
+                    PreferenceUtils.setVippsToken(context.getApplicationContext(), response);
+                    if (response.isEmpty()) {
+                        Log.e(TAG, "Got response from server, but token was empty");
+                    } else {
+                        Log.i(TAG, "Successful fetch");
+                    }
+                }, error -> {
+            cached.setValue(Resource.error(String.valueOf(error.networkResponse.statusCode), error));
+            Log.e(TAG, (String.valueOf(error.networkResponse.statusCode)));
+        }) {
+            @Override
+            public Map<String, String> getHeaders() {
+                Map<String, String> map = new HashMap<>();
+                map.put(CLIENT_ID, CommunicationConfig.getInstance().retrieveClientID());
+                map.put(CLIENT_SECRET, CommunicationConfig.getInstance().retrieveClientSecret());
+                map.put(OCP_APIM_SUBSCRIPTION_KEY, CommunicationConfig.getInstance().retrieveOcpApimSubscriptionKey());
+                return map;
+            }
+        };
+        requestQueue.add(request);
+    }
```

Comment:
Use cached instead of prefs at all?
if (agree) {
createNewIssue("use cache");
} else {
killSelf();
}